util/cpuprofile/cpuprofile.go

```diff
@@ -0,0 +1,243 @@
+		cs:     make(map[ProfileConsumer]struct{}),
+		notify: make(chan struct{}),
+		closed: make(chan struct{}),
+	}
+}
+
+var (
+	errProfilerAlreadyStarted = errors.New("parallelCPUProfiler already started")
+	errProfilerAlreadyClosed  = errors.New("parallelCPUProfiler already closed")
+)
+
+func (p *parallelCPUProfiler) start() error {
+	if !p.started.CAS(false, true) {
+		return errProfilerAlreadyStarted
+	}
+	if p.isClosed.Load() {
+		return errProfilerAlreadyClosed
+	}
+
+	logutil.BgLogger().Info("parallel cpu profiler started")
+	p.wg.Add(1)
+	go util.WithRecovery(p.profilingLoop, nil)
+	return nil
+}
+
+func (p *parallelCPUProfiler) close() {
+	if p.isClosed.CAS(false, true) {
+		close(p.closed)
+	}
+	p.wg.Wait()
+}
+
+func (p *parallelCPUProfiler) register(ch ProfileConsumer) {
+	if ch == nil {
+		return
+	}
+	p.Lock()
+	p.cs[ch] = struct{}{}
+	p.Unlock()
+	// notify
+	select {
+	case p.notify <- struct{}{}:
+	default:
+	}
+}
+
+func (p *parallelCPUProfiler) unregister(ch ProfileConsumer) {
+	if ch == nil {
+		return
+	}
+	p.Lock()
+	delete(p.cs, ch)
+	p.Unlock()
+}
+
+func (p *parallelCPUProfiler) profilingLoop() {
+	checkTicker := time.NewTicker(defProfileDuration)
+	defer func() {
+		checkTicker.Stop()
+		pprof.StopCPUProfile()
+		p.wg.Done()
+	}()
+	for {
+		select {
+		case <-p.closed:
+			return
+		case <-p.notify:
+			if !p.inProfilingStatus() {
+				p.doProfiling()
+			}
+		case <-checkTicker.C:
+			p.doProfiling()
+		}
+	}
+}
+
+func (p *parallelCPUProfiler) doProfiling() {
+	if p.inProfilingStatus() {
+		p.stopCPUProfile()
+		p.sendToConsumers()
+	}
+
+	// Only do cpu profiling when there are consumers.
+	if p.needProfile() {
+		err := p.startCPUProfile()
+		if err != nil {
+			p.data.Error = err
+			// notify error as soon as possible
+			p.sendToConsumers()
+		}
+	}
+}
+
+func (p *parallelCPUProfiler) needProfile() bool {
+	return p.consumersCount() > 0
+}
+
+func (p *parallelCPUProfiler) inProfilingStatus() bool {
+	return p.data != nil
+}
+
+func (p *parallelCPUProfiler) startCPUProfile() error {
+	if p.nextData != nil {
+		p.data, p.nextData = p.nextData, nil
+	} else {
+		p.data = &ProfileData{Data: bytes.NewBuffer(make([]byte, 0, 1024*8)), Begin: time.Now()}
+	}
+	metrics.CPUProfileCounter.Inc()
+	return pprof.StartCPUProfile(p.data.Data)
+}
+
+func (p *parallelCPUProfiler) stopCPUProfile() {
+	now := time.Now()
+	p.data.End = now
+
+	// create next profile data before stop profiling, to reduce profiling interval.
+	if p.lastDataSize != 0 {
+		p.nextData = &ProfileData{Data: bytes.NewBuffer(make([]byte, 0, p.lastDataSize)), Begin: now}
+	}
+
+	pprof.StopCPUProfile()
+}
+
+func (p *parallelCPUProfiler) consumersCount() int {
+	p.Lock()
+	n := len(p.cs)
+	p.Unlock()
+	return n
+}
+
+func (p *parallelCPUProfiler) sendToConsumers() {
+	p.Lock()
+	for c := range p.cs {
+		select {
+		case c <- p.data:
+		default:
+			// ignore
+		}
+	}
+	p.Unlock()
+	p.lastDataSize = p.data.Data.Len()
+	p.data = nil
+}
```

Comment:
Please reduce your L82 to L243 to within 50 LOC. You made a simple thing really unnecessarily complex.